trieste/models/model_interfaces.py

```diff
@@ -94,6 +94,95 @@ def optimize(self, dataset: Dataset) -> None:
+            method signature requires at least one model. It is not treated specially.
+        :param \*models_with_event_sizes: The other models, and sizes of their output events.
+        """
+        super().__init__()
+        self._models, self._event_sizes = zip(*(model_with_event_size,) + models_with_event_sizes)
+
+    def predict(self, query_points: TensorType) -> tuple[TensorType, TensorType]:
+        r"""
+        :param query_points: The points at which to make predictions, of shape [..., D].
+        :return: The predictions from all the wrapped models, concatenated along the event axis in
+            the same order as they appear in :meth:`__init__`. If the wrapped models have predictive
+            distributions with event shapes [:math:`E_i`], the mean and variance will both have
+            shape [..., :math:`\sum_i E_i`].
+        """
+        means, vars_ = zip(*[model.predict(query_points) for model in self._models])
+        return tf.concat(means, axis=-1), tf.concat(vars_, axis=-1)
+
+    def predict_joint(self, query_points: TensorType) -> tuple[TensorType, TensorType]:
+        r"""
+        :param query_points: The points at which to make predictions, of shape [..., B, D].
+        :return: The predictions from all the wrapped models, concatenated along the event axis in
+            the same order as they appear in :meth:`__init__`. If the wrapped models have predictive
+            distributions with event shapes [:math:`E_i`], the mean will have shape
+            [..., B, :math:`\sum_i E_i`], and the covariance shape
+            [..., :math:`\sum_i E_i`, B, B].
+        """
+        means, covs = zip(*[model.predict_joint(query_points) for model in self._models])
+        return tf.concat(means, axis=-1), tf.concat(covs, axis=-3)
+
+    def sample(self, query_points: TensorType, num_samples: int) -> TensorType:
+        r"""
+        :param query_points: The points at which to sample, with shape [..., D].
+        :param num_samples: The number of samples at each point.
+        :return: The samples from all the wrapped models, concatenated along the event axis. For
+            wrapped models with predictive distributions with event shapes [:math:`E_i`], this has
+            shape [..., S, :math:`\sum_i E_i`], where S is the number of samples.
+        """
+        samples = [model.sample(query_points, num_samples) for model in self._models]
+        return tf.concat(samples, axis=-1)
+
+    def update(self, dataset: Dataset) -> None:
+        """
+        Update all the wrapped models on their corresponding data. The data for each model is
+        extracted by splitting the observations in ``dataset`` along the event axis according to the
+        event sizes specified at :meth:`__init__`.
+
+        :param dataset: The query points and observations for *all* the wrapped models.
+        """
+        observations = tf.split(dataset.observations, self._event_sizes, axis=-1)
```

Comment:
I'm trying to understand what the dataset will look like here.

 So dataset.observations will have an additional dimension that corresponds to different events. So we have observations for each individual event? What exactly is an event here?

Reply:
An event is a subset of the observations. For example, if I have 3 objectives, and I have a GPR for the first and SVGP for the last two, an event could be an observation of the first objective, or an observation of the last two objectives. The event sizes would be 1 and 2.

Reply:
I've used event in other places in trieste to mean all the observations, and sometimes to mean a query point. Mainly I use it to mean a set of observations which are not independent.